packages/alfa-painting-order/src/painting-order.ts

```diff
@@ -0,0 +1,286 @@
+import { Array } from "@siteimprove/alfa-array";
+import { Cache } from "@siteimprove/alfa-cache";
+import { Comparable } from "@siteimprove/alfa-comparable";
+import type { Device } from "@siteimprove/alfa-device";
+import { Element, Node } from "@siteimprove/alfa-dom";
+import type { Equatable } from "@siteimprove/alfa-equatable";
+import type { Hash, Hashable } from "@siteimprove/alfa-hash";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import type { Serializable } from "@siteimprove/alfa-json";
+import { Refinement } from "@siteimprove/alfa-refinement";
+import { Style } from "@siteimprove/alfa-style";
+
+import * as json from "@siteimprove/alfa-json";
+
+const { and, not, or } = Refinement;
+const {
+  hasInitialComputedStyle,
+  isBlockContainer,
+  isFlexOrGridChild,
+  isPositioned,
+  isRendered,
+} = Style;
+
+import { Sequence } from "@siteimprove/alfa-sequence";
+import { createsStackingContext } from "./predicate/creates-stacking-context.js";
+
+/**
+ * @public
+ */
+export class PaintingOrder
+  implements Equatable, Hashable, Serializable<PaintingOrder.JSON>
+{
+  public static of(elements: Iterable<Element>): PaintingOrder {
+    return new PaintingOrder(Array.from(elements));
+  }
+
+  private readonly _elements: Array<Element>;
+
+  protected constructor(elements: Array<Element>) {
+    this._elements = elements;
+  }
+
+  public equals(value: this): boolean;
+  public equals(value: unknown): value is this;
+  public equals(value: unknown): boolean {
+    return (
+      value === this ||
```

Comment:
```suggestion
```
We usually don't have the physical equality test in our `.equals` (which does indeed make them less efficient in that case, but that is hopefully a rare case 🤔